keps/sig-cluster-lifecycle/0015-kubeadm-join-master.md

```diff
@@ -0,0 +1,431 @@
+Given such information, the user can choose a different approach where each master has a
+specific apiserver advertise address different from the `controlplaneAddress`.
+
+Please note that:
+
+- if the user is not planning to distribute the apiserver certificate among masters, kubeadm
+  will generate a new apiserver serving certificate with the required SANS
+- if the user is planning to distribute the apiserver certificate among masters, he/she/the
+  operator is required to provide during `kubeadm init` the list of masters/the list of IP
+  addresses for all the masters as alternative names for the API servers certificate, thus
+  allowing the proper functioning of all the API server instances that will join
+
+#### Strategies for deploying control plane components
+
+As of today kubeadm supports two solutions for deploying control plane components:
+
+1. Control plane deployed as static pods (current kubeadm default)
+2. Self-hosted control plane in case of `--feature-gates=SelfHosting=true`
+
+The proposed solution for case 1. "Control plane deployed as static pods", assumes
+that the `kubeadm join --master` flow will take care of creating required kubeconfig
+files and required static pod manifests.
+
+Self-hosted control plane are not supported in this release of the proposal
+(`kubeadm join --master` will fail immediately), but nothing in this proposal should
+prevent to address this in subsequent phases.
+
+#### Strategies for distributing cluster certificates
+
+As of today kubeadm supports two solutions for storing cluster certificates:
+
+1. Cluster certificates stored on file system in case of:
+   - Control plane deployed as static pods (current kubeadm default)
+   - Self-hosted control plane in case of `--feature-gates=SelfHosting=true,StoreCertsInSecrets=false`
+2. Cluster certificates stored in secrets in case of:
+   - Self-hosted control plane + secrets in certs in case of
+     `--feature-gates=SelfHosting=true,StoreCertsInSecrets=true`
+
+The proposed solution for case 1. "Cluster certificates stored on file system",
+requires the user/the higher level tools to execute an additional action _before_
+invoking `kubeadm join --master`.
+
+More specifically, in case of cluster with "cluster certificates stored on file
+system", before invoking `kubeadm join --master`, the user/higher level tools
+should copy control plane certificates from an existing master node, e.g. bootstrap master
+
+> NB. kubeadm is limited to execute actions *only*
+in the machine where it is running, so it is not possible to copy automatically
+certificates from remote locations.
+
+Then, the `kubeadm join --master` flow will take care of checking certificates
+existence and conformance.
+
+"Cluster certificates stored in secrets", are not supported in this release of the
+proposal (`kubeadm join --master` will fail immediately), but nothing in this proposal
+should prevent to address this in subsequent phases.
+
+#### `kubeadm upgrade` for HA clusters
```

Comment:
can you dive into exactly _how_ and in which order the upgrade should be performed.
i.e. upgrade etcd fully first (in _some_ way out of scope for this proposal probably), then remove one master from the loadbalancer, upgrade it, (add it back to the LB, or not). Upgrades without downtime supported officially or not? etc. etc.

Reply:
The only prior art I'm aware of is https://kubernetes.io/docs/tasks/administer-cluster/kubeadm/kubeadm-upgrade-ha/ , but I'm not sure there is already a consensus about how to do this, and so I prefer to get a first release of the KEP approved and the iterate on the upgrade part.